src/internal/storage/chunk/transform.go

```diff
@@ -47,33 +50,38 @@ func Create(ctx context.Context, opts CreateOptions, ptext []byte, createFunc fu
 
 // Get calls getFunc to retrieve a chunk, then verifies, decrypts, and decompresses the data.
 // cb is called with the uncompressed plaintext
-func Get(ctx context.Context, client Client, cache kv.GetPut, ref *Ref, cb kv.ValueCallback) error {
-	if err := getFromCache(ctx, cache, ref, cb); err == nil {
-		return nil
-	}
+func Get(ctx context.Context, client Client, cache kv.GetPut, deduper *miscutil.WorkDeduper, ref *Ref, cb kv.ValueCallback) error {
 	if ref.EncryptionAlgo != EncryptionAlgo_CHACHA20 {
 		return errors.Errorf("unknown encryption algorithm %d", ref.EncryptionAlgo)
 	}
-	return client.Get(ctx, ref.Id, func(ctext []byte) error {
-		if err := verifyData(ref.Id, ctext); err != nil {
-			return err
-		}
-		var r io.Reader = bytes.NewReader(ctext)
-		var err error
-		if r, err = decrypt(ref.Dek, r); err != nil {
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = 1 * time.Millisecond
+	return backoff.RetryUntilCancel(ctx, func() error {
+		return getFromCache(ctx, cache, ref, cb)
+	}, b, func(err error, _ time.Duration) error {
```

Comment:
This is definitely not what the author of backoff intended.  Literally everything is happening in a callback that's supposed to be for logging.

The logic should be (in a loop).
1. check the cache, if it's there then call the callback.
2. fill the cache.  This is done in the deduper.  And will call the transform Get with a callback.  The transform Get loses knowledge of the cache.

After like 10 tries fallback to just getting it manually.  Alternatively you can call cb inside the deduper, but that will block other calls, which could use it if it were just in the cache.

Reply:
Yeah, so this probably wasn't the intended use of the notify callback, but I think it is preferable to what you suggested because it is simpler and includes an exponential backoff. I don't think there is any reason the notify callback cannot be more general purpose than just logging.

Reply:
You could just move this into the regular callback.  It would do the same thing.